src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,664 @@
+ * {@linkplain java.util.concurrent##MemoryVisibility <em>happens-before</em>}
+ * any subsequent read operation (e.g. {@linkplain #orElseThrow()}).
+ * <p>
+ * The method {@linkplain StableValue#orElseSet(Supplier) orElseSet()} guarantees that
+ * the provided {@linkplain Supplier} is invoked successfully at most once even under
+ * race. Since stable functions and stable collections are built on top of
+ * {@linkplain StableValue#orElseSet(Supplier) orElseSet()} they too are
+ * thread safe and guarantee at-most-once-per-input invocation.
+ *
+ * @implSpec Implementing classes of {@linkplain StableValue} are free to synchronize on
+ *           {@code this} and consequently, care should be taken whenever
+ *           (directly or indirectly) synchronizing on a {@code StableValue}. Failure to
+ *           do this may lead to deadlock. Stable functions and collections on the
+ *           other hand are guaranteed <em>not to synchronize</em> on {@code this}.
+ *           Except for a {@code StableValue}'s content itself, an {@linkplain #orElse(Object) orElse(other)}
+ *           parameter, and an {@linkplain #equals(Object) equals(obj)} parameter; all
+ *           method parameters must be <em>non-null</em> or a {@link NullPointerException}
+ *           will be thrown.
+ *
+ * @implNote A {@linkplain StableValue} is mainly intended to be a non-public field in
+ *           a class and is usually neither exposed directly via accessors nor passed as
+ *           a method parameter.
+ *           As objects can be set via stable values but never removed, this can be a source
+ *           of unintended memory leaks. A stable value's content is
+ *           {@linkplain java.lang.ref##reachability strongly reachable}. Clients are
+ *           advised that {@linkplain java.lang.ref##reachability reachable} stable values
+ *           will hold their set content perpetually.
+ *           A {@linkplain StableValue} that has a type parameter {@code T} that is an array
+ *           type (of arbitrary rank) will only allow the JVM to treat the <em>array reference</em>
+ *           as a stable value but <em>not its components</em>. Clients can instead use
+ *           {@linkplain #list(int, IntFunction) a stable list} of arbitrary depth, which
+ *           provides stable components. More generally, a stable value can hold other
+ *           stable values of arbitrary depth and still provide transitive constantness.
+ *
+ * @param <T> type of the content
+ *
+ * @since 25
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.STABLE_VALUES)
+public sealed interface StableValue<T>
+        permits StableValueImpl {
+
+    // Principal methods
+
+    /**
+     * {@return {@code true} if the content was set to the provided {@code value},
+     *          {@code false} otherwise}
+     * <p>
+     * When this method returns, the content is always set.
+     *
+     * @param value to set
+     */
+    boolean trySet(T value);
```

Comment:
This can probably throw `IllegalStateException` too right -- e.g. if used inside a supplier passed to `orElseSet` ?

Reply:
Nice catch.